src/libinjector/linux/methods/linux_shellcode.c

```diff
@@ -0,0 +1,267 @@
+            PRINT_DEBUG("memory address allocated using mmap: %lx\n", info->regs->rax);
+
+            // save it for future use
+            injector->virtual_memory_addr = info->regs->rax;
+
+            if (!write_shellcode_to_mmap_location(drakvuf, info))
+                return cleanup(drakvuf, info, true);
+
+            setup_mmap_trap(drakvuf, info);
+            info->regs->rip = injector->virtual_memory_addr;
+
+            free_bp_trap(drakvuf, injector, info->trap);
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP3: //since mmap starting location is trapped, the first one will be this
+        {
+            PRINT_DEBUG("Shellcode begin\n");
+
+            if (!save_rip_for_ret(drakvuf, info->regs))
+                return cleanup(drakvuf, info, true);
+
+            // rsp is being updated
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP4: // shellcode should've executed and we will be returned to the same trap as ret is appended in the end
+        {
+            PRINT_DEBUG("Shellcode end\n");
+            free_bp_trap(drakvuf, injector, info->trap);
+
+            // restore regs
+            memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t));
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP5: // cleanup
+        {
+            PRINT_DEBUG("Removing traps and exiting\n");
+
+            // remove the initial trap here
+            free_bp_trap(drakvuf, injector, info->trap);
+            drakvuf_interrupt(drakvuf, SIGINT);
+
+            event = VMI_EVENT_RESPONSE_NONE;
+            break;
+        }
+        default:
+        {
+            PRINT_DEBUG("Should not be here\n");
+            assert(false);
+        }
+    }
+
+    return event;
+}
+
+static event_response_t cleanup(drakvuf_t drakvuf, drakvuf_trap_info_t* info, bool clear_trap)
+{
+    fprintf(stderr, "Doing premature cleanup\n");
+    injector_t injector = (injector_t)info->trap->data;
+
+    // restore regs
+    memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t));
+
+    if (clear_trap)
+        free_bp_trap(drakvuf, injector, info->trap);
+
+    // since we are jumping to some arbitrary step, we will set this
+    injector->step_override = true;
+
+    // give the last step
+    injector->step = STEP5;
+
+    return VMI_EVENT_RESPONSE_SET_REGISTERS;
+}
+
+static bool setup_mmap_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    injector_t injector = info->trap->data;
+
+    injector->bp = g_try_malloc0(sizeof(drakvuf_trap_t));
+
+    injector->bp->type = BREAKPOINT;
+    injector->bp->name = "injector_mmap_trap";
+    // cb will be set from previous call only
+    // we don't have injector_int3_userspace_cb function
+    // in scope here so we will use it from the previous trap
+    injector->bp->cb = info->trap->cb; //injector_int3_userspace_cb;
```

Comment:
maybe remove `static` from `injector_int3_userspace_cb` if it's required in other parts of the code?

Reply:
umm, could you give a proper way in which I can do that? ( coz `linux_injector.c` doesn't have any header files, and also, it is including `linux_shellcode.c` and `linux_utils.c` which will make a recursive header file inclusion even if I make some header file, so am not sure on how to do it )

Reply:
then create some new headers :)

Reply:
Done